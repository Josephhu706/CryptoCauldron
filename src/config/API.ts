import axios from "axios";

export type Coin = {
    id: string;
    symbol: string;
    name: string;
    image: string;
    current_price: number;
    market_cap: number;
    market_cap_rank: number;
    fully_diluted_valuation: number;
    total_volume: number;
    high_24h: number;
    low_24h: number;
    price_change_24h: number;
    price_change_percentage_24h: number;
    market_cap_change_24h: number;
    market_cap_change_percentage_24h: number;
    circulating_supply: number;
    total_supply: number;
    max_supply: number;
    ath: number;
    ath_change_percentage: number;
    ath_date: string;
    atl: number;
    atl_change_percentage: number;
    atl_date: string;
    roi: any;
    last_updated: string;
}

export const CoinList = () =>
`https://api.coingecko.com/api/v3/coins/markets?vs_currency=aud&order=market_cap_desc&per_page=100&page=1&sparkline=false`;

export const SingleCoin = (id:number) =>
`https://api.coingecko.com/api/v3/coins/${id}`;

export const HistoricalChart = (id:number, days = 365) =>
`https://api.coingecko.com/api/v3/coins/${id}/market_chart?vs_currency=aud&days=${days}`;

export const TrendingCoins = () =>
`https://api.coingecko.com/api/v3/coins/markets?vs_currency=aud&order=gecko_desc&per_page=10&page=1&sparkline=false&price_change_percentage=24h`;

export const fetchCrypto = async ()=>{
    //put page not found 
    const endpoint = `https://api.coingecko.com/api/v3/coins/markets?vs_currency=aud&order=market_cap_desc&per_page=100&page=1&sparkline=false`;
    const {data} = await axios.get(endpoint)
    console.log(data)
    return data

}