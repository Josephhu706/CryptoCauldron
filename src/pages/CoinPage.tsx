import {useParams} from 'react-router-dom'
import React, {useEffect, useState} from 'react';
import {fetchCoin} from '../config/API'
import { ThemeProvider, createTheme } from '@mui/material/styles';
import { CssBaseline } from '@mui/material/';
import CoinChart from '../components/CoinChart';
import CircularProgress from '@mui/material/CircularProgress'
import '../styles/CoinPageStyles.scss'
import callError from '../assets/images/call-error.svg'
import Typography from '@mui/material/Typography';
import parse from 'html-react-parser';
import Button from '@mui/material/Button';
import ArrowBackIosIcon from '@mui/icons-material/ArrowBackIos';
import {useNavigate} from "react-router-dom";

const darkTheme = createTheme({
  palette: {
    mode: 'dark',
  },
});

const CoinPage = () => {
  const {id} = useParams()
  const [coin, setCoin] =useState<any>()
  const [loading, setLoading] = useState(false)
  const [error, setError] = useState<string>('')

  const navigate = useNavigate();

  useEffect(() => {
    getCoin(id)
  }, [])

  const getCoin = async (id:any)=>{
    setLoading(true)
    try{
      const newCoin = await fetchCoin(id)
      setCoin(newCoin)
    }catch(error:any){
      setError(error)
    }
    setLoading(false)
  }



  return (
    <>
      <ThemeProvider theme={darkTheme}>
          <CssBaseline/>
          {loading? (
                  <CircularProgress data-testid='loading' />        
                ):(error? 
                <div data-testid='fetchError' className="fetchError">
                  <img src={callError} style={{"height": "20em"}} alt="fetchError"/>
                </div>
                :(
                  <div data-testid='coinPage' className='coinPageContainer'>

                    <div className="sideBar">
                    <Button className="backButton" variant="outlined" onClick={()=>navigate('/')}>
                          <ArrowBackIosIcon/> Back To Home
                    </Button>
                    
                      <img className='coinImage' src={coin?.image.large} alt={coin?.name} />
                      <Typography variant='h3' className="headingName">
                        {coin?.name}
                      </Typography>
                      <Typography variant='subtitle1' className="coinDescription">
                        {parse(`${coin?.description.en.split(". ")[0]}`)}
                      </Typography>
                      <div className="marketData">
                        <span className="coinTag coinRank">
                          <Typography variant="h5" className="headingName">Rank: <span className="marketDataContent">{coin?.market_cap_rank}</span></Typography>
                        </span>
                        <span className="coinTag coinPrice">
                          <Typography variant="h5" className="headingName">Price: <span className="marketDataContent">${coin?.market_data.current_price.aud.toLocaleString()}</span></Typography>
                        </span>
                        <span className="coinTag coinCap">
                          <Typography variant="h5" className="headingName">Market Cap: <span className="marketDataContent">${coin?.market_data.market_cap.aud.toLocaleString()}</span></Typography>
                        </span>
                      </div>
                    </div>


                    <div className="coinChart">
                      <CoinChart coin={coin?.id}/>
                    </div>

                </div>
                )
                )}
      </ThemeProvider>
    </>
  )
}

export default CoinPage