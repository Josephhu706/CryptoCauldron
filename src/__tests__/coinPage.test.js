import { render, screen, cleanup, getByTestId, fireEvent, waitFor, getByText, act} from '@testing-library/react'
import CoinPage from '../pages/CoinPage'
import '@testing-library/jest-dom'
import axiosMock from 'axios'
import renderer from 'react-test-renderer'
import {fetchCoin, fetchHistory} from '../config/API'


afterEach(()=>{
    cleanup()
})

jest.mock('axios');
jest.mock('../config/API')

const mockedUsedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
   ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUsedNavigate,
}));

test("Coin Page component should be rendered", () => {
    act(()=>render(<CoinPage/>))
    const coinPage = screen.getByTestId("coinPage");
    expect(coinPage).toBeInTheDocument()
});


test("Render Error Message if Fetch Fails", async () => {
    axiosMock.get.mockRejectedValue({data: []})
    act(()=>render(<CoinPage/>))
    const loading= screen.getByTestId("loading")
    expect(loading).toBeInTheDocument();
    expect(fetchCoin).toHaveBeenCalledTimes(1)
    await waitFor(()=>{
      const errorMsg = screen.getByTestId('fetchError')
      expect(errorMsg).toBeInTheDocument();
    })
});

test("Render coin information", async () => {
    axiosMock.get.mockResolvedValueOnce({data: {name: "Bitcoin", id:"bitcoin"}})
    act(()=>render(<CoinPage/>))
    const loading= screen.getByTestId("loading")
    expect(loading).toBeInTheDocument();
    expect(fetchCoin).toHaveBeenCalledTimes(1)
    waitFor(()=>{
        const coinContent = screen.findByText('Bitcoin')
        const coinName = screen.findByTestId('coinName')
        expect(coinName).toBeInTheDocument()
        expect(coinContent).toBeInTheDocument()
    })
});

test("Render Home Button and navigate back Home", async () => {
    axiosMock.get.mockResolvedValueOnce({data: {name: "Bitcoin", id:"bitcoin"}})
    act(()=>render(<CoinPage/>))
    const loading= screen.getByTestId("loading")
    expect(loading).toBeInTheDocument();
    expect(fetchCoin).toHaveBeenCalledTimes(1)
    waitFor(()=>{
        const homeButton = screen.findByText('BACK TO HOME')
        expect(homeButton).toBeInTheDocument()
        fireEvent.click(homeButton)
        expect(mockedUsedNavigate).toHaveBeenCalledWith(`/`)
    })
});

  
test('Coin page matches snapshot', () =>{
    const tree = renderer.create(<CoinPage/>).toJSON()
    expect(tree).toMatchSnapshot();

})



