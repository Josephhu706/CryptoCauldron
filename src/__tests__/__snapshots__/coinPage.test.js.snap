// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Coin page matches snapshot 1`] = `
<div
  data-testid="coinPage"
>
  <div
    className="coinPageContainer"
  >
    <div
      className="sideBar"
      data-testid="sideBar"
    >
      <button
        className="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root backButton css-1vbvugf-MuiButtonBase-root-MuiButton-root"
        data-testid="backButton"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ArrowBackIosIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M11.67 3.87 9.9 2.1 0 12l9.9 9.9 1.77-1.77L3.54 12z"
          />
        </svg>
         Back To Home
      </button>
      <img
        className="coinImage"
      />
      <h3
        className="MuiTypography-root MuiTypography-h3 headingName css-gepadz-MuiTypography-root"
        data-testid="coinName"
      />
      <h6
        className="MuiTypography-root MuiTypography-subtitle1 coinDescription css-10wpov9-MuiTypography-root"
      />
      <div
        className="marketData"
      >
        <span
          className="coinTag coinRank"
        >
          <h5
            className="MuiTypography-root MuiTypography-h5 headingName css-ag7rrr-MuiTypography-root"
          >
            Rank:
             
            <span
              className="marketDataContent"
            />
          </h5>
        </span>
        <span
          className="coinTag coinPrice"
        >
          <h5
            className="MuiTypography-root MuiTypography-h5 headingName css-ag7rrr-MuiTypography-root"
          >
            Price:
             
            <span
              className="marketDataContent"
            >
              $
            </span>
          </h5>
        </span>
        <span
          className="coinTag coinCap"
        >
          <h5
            className="MuiTypography-root MuiTypography-h5 headingName css-ag7rrr-MuiTypography-root"
          >
            Market Cap:
             
            <span
              className="marketDataContent"
            >
              $
            </span>
          </h5>
        </span>
      </div>
    </div>
    <div
      className="coinChart"
    >
      <div
        className="container"
      >
        <h1
          className="coinHeading"
        >
          Coin Price Data
        </h1>
        <div
          className="coinChart"
          data-testid="coinChart"
        >
          <canvas
            height={150}
            role="img"
            width={300}
          />
        </div>
        <div
          className="coinChartButtons"
        >
          <span>
            <button
              className="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root chartButton css-1vbvugf-MuiButtonBase-root-MuiButton-root"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              24 Hours
            </button>
          </span>
          <span>
            <button
              className="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root chartButton css-1vbvugf-MuiButtonBase-root-MuiButton-root"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              30 Days
            </button>
          </span>
          <span>
            <button
              className="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root chartButton css-1vbvugf-MuiButtonBase-root-MuiButton-root"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              3 Months
            </button>
          </span>
          <span>
            <button
              className="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root chartButton css-1vbvugf-MuiButtonBase-root-MuiButton-root"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              1 Year
            </button>
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
